File: src/apps/services/src/controllers/user/user.controller.ts
import * as tedious from 'tedious';
import { Request, Response } from 'express';
import IUser from '../../models/user.interface';
import { connection as conn } from '../../main';
export default class UserController {
  public getAllUsers = (req: Request, res: Response) => {
    const query = 'SELECT * FROM [dbo].[users]';
    const users: IUser[] = [];

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('Express: ' + err.message);
            return res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            console.log('Express: No users exist.');
            return res.status(404).json({
              error: 'Not Found',
              details: 'No users exist.',
            });
          } else {
            console.log('Express: users retrieved successfully.');
            res.status(200).json(users);
          }
        }
      );

      request.on('row', (columns: tedious.ColumnValue[]) => {
        const user: IUser = {
          userId: columns[0].value,
          email: columns[1].value,
          password: columns[2].value,
          userRole: columns[3].value,
          dateCreated: columns[4].value,
          lastLoggedIn: columns[5].value,
        };
        users.push(user);
      });

      conn.execSql(request);
    } catch (error) {
      console.log('Express: ' + error.message);
      res.status(500).json({
        error: error.message,
      });
    }
  };

  public getUser = (req: Request, res: Response) => {
    const { userId } = req.params;
    let user: IUser;
    const query = `SELECT * FROM [dbo].[users] WHERE userId = ${userId}`;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('Express: ' + err.message);
            return res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            console.log('Express: User does not exist.');
            return res.status(404).json({
              error: 'Not Found',
              details: 'User does not exist.',
            });
          } else {
            console.log('Express: user retrieved successfully.');
            res.status(200).json(user);
          }
        }
      );

      request.on('row', (columns: tedious.ColumnValue[]) => {
        user = {
          userId: columns[0].value,
          email: columns[1].value,
          password: columns[2].value,
          userRole: columns[3].value,
          dateCreated: columns[4].value,
          lastLoggedIn: columns[5].value,
        };
      });

      conn.execSql(request);
    } catch (error) {
      console.log('Express: ' + error.message);
      res.status(500).json({
        error: error.message,
      });
    }
  };

  public updateUser = (req: Request, res: Response) => {
    const { userId } = req.params;
    const { email, password, userRole } = req.body;
    const currentDate = new Date().toISOString().slice(0, 19).replace('T', ' ');
    const query = `UPDATE [dbo].[users] SET email = '${email}', password = '${password}', userRole = ${userRole}, lastLoggedIn = '${currentDate}' WHERE userId = ${userId}`;
    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('Express: ' + err.message);
            return res.status(404).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            console.log('Express: User does not exist.');
            return res.status(404).json({
              error: 'Not Found',
              details: 'User does not exist.',
            });
          } else {
            console.log('Express: user updated successfully.');
            res.status(200).json({
              message: 'User updated successfully.',
            });
          }
        }
      );

      conn.execSql(request);
    } catch (error) {
      console.log('Express: ' + error.message);
      res.status(500).json({
        error: error.message,
      });
    }
  };

  public deleteUser = (req: Request, res: Response) => {
    const { userId } = req.params;
    const query = `DELETE FROM [dbo].[users] WHERE userId = ${userId}`;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('Express: ' + err.message);
            return res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            console.log('Express: User does not exist.');
            return res.status(404).json({
              error: 'Not Found',
              details: 'User does not exist.',
            });
          } else {
            console.log('Express: user deleted successfully.');
            res.status(200).json({
              message: 'User deleted successfully.',
            });
          }
        }
      );

      conn.execSql(request);
    } catch (error) {
      console.log('Express: ' + error.message);
      res.status(500).json({
        error: error.message,
      });
    }
  };
}

File: src/apps/api/Controllers/UserController.cs
using Microsoft.AspNetCore.Mvc;
using Api.Repository;

namespace Api.Controllers;

[ApiController]
[Route("[controller]")]
public class UserController : ControllerBase
{
    private readonly UsersRepository _usersRepository;

    public UserController()
    {
        _usersRepository = new UsersRepository();
    }

    [HttpGet]
    [Route("all")]
    public async Task<IActionResult> GetAllUsers()
    {
        try
        {
            var data = await _usersRepository.GetAllUsers();
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

    //Update an user
    [HttpPatch]
    [Route("update")]
    public async Task<IActionResult> UpdateUser([FromBody] Users user)
    {
        try
        {
            var data = await _usersRepository.UpdateUsers(
                user.userId,
                user.email!,
                user.password!,
                user.userRole,
                user.dateCreated,
                user.lastLoggedIn
            );
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

    //Delete an user
    [HttpDelete]
    [Route("delete")]
    public async Task<IActionResult> DeleteUser([FromBody] Users user)
    {
        try
        {
            var data = await _usersRepository.DeleteUsers(user.userId);
            if (data == false)
            {
                return StatusCode(404, "User with id: " + user.userId + " does not exist");
            }
            return Ok("Deleted user with id: " + user.userId + "");
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

    //Get an user by id
    [HttpGet]
    [Route("get/{id}")]
    public async Task<IActionResult> GetUserById([FromRoute] int id)
    {
        try
        {
            var data = await _usersRepository.GetUserById(id);
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(404, e.Message);
        }
    }
}
